### PR TITLE
Unique-10 operator specification

### PR DESCRIPTION
**Pull request by tomdol**
### Details:
- The PR contains the specification of the Unique op candidate. The spec covers both TF and ONNX use cases.
- ONNX spec: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Unique
- TF spec: https://www.tensorflow.org/api_docs/python/tf/unique
- `axis` as an input instead of an attribute (ONNX approach) to make it possible to NOT specify the axis at all
- the element type of the `axis` input limited to i32 and i64 only - do we need a broader range of types? The `axis` is only required by ONNX model for now and in those models this value is an integer attribute that we need to transform into an OV Constant during the import phase.
- type of the T_IND configurable with an attribute called 'index_element_type' - ONNX always uses i64 to return the indices but TF allows i32 and i64. Should we support unsigned types here?
- the last output tensor only allows int64 to match the ONNX spec. Should we extend the number of allowed types here? Should it be configurable just like the `index_element_type` is?

### Tickets:
 - *91518*

### CODE REVIEW

File: docs/ops/movement/Unique_10.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+**Attributes**:
+
+* *axis*
+
+  * **Description**: *axis* specifies the input tensor's axis along which the subtensors are processed. When the axis is set to the maximum value of the underlying integer type it means that the input tensor should be processed elementwise (the operator then looks for unique individual numbers in the whole flattened tensor).
+  * **Range of values**: The accepted range of values is `[-R, R-1]` where `R` is the rank of the input tensor.
+  * **Type**: int
+  * **Default value**: MAX_INT
+  * **Required**: *no*
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:20]: Why do we have this attribute if we have dedicated input for this?

File: docs/ops/movement/Unique_10.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+* **2**: The output tensor containing indices of the locations of unique elements. The indices map the elements in the first output tensor to their locations in the input tensor. The index always points to the first occurence of a given unique output element in the input tensor. This is a 1D tensor with type *T_IND*.
+* **3**: The output tensor containing indices of the locations of elements of the input tensor in the first output tensor. This means that for each element of the input tensor this output will point to the unique value in the first output tensor of this operator. This is a 1D tensor with type *T_IND*.
+* **4**: The output tensor containing the number of occurences of each unique value produced by this operator in the first output tensor. This is a 1D tensor with type `int64`.
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:51]: What type does PyTorch return for this output? Could you please check?

File: docs/ops/movement/Unique_10.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+* *sorted*
+
+  * **Description**: controls whether the unique elements in the output tensor are sorted in ascending order.
+  * **Range of values**:
+    * false - output tensor's elements are not sorted
+    * true - output tensor's element are sorted
+  * **Type**: boolean
+  * **Default value**: true
+  * **Required**: *no*
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:30]: How are they ordered in case of TensorFlow and PyTorch without sort. Are they aligned with ONNX or between each other?

File: docs/ops/movement/Unique_10.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
+# Unique  {#openvino_docs_ops_comparison_Unique_10}
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:1]: I think you also need to create opset10.md file and add reference to opset10 from the content.

File: docs/ops/movement/Unique_10.md (hunk 1 above)
Comment (tomdol) [reply docs/ops/movement/Unique_10.md:20]: This an oversight, thanks for catching it

File: docs/ops/movement/Unique_10.md (hunk 2 above)
Comment (tomdol) [reply docs/ops/movement/Unique_10.md:51]: Sure, I've checked in manually in python because this information is missing in the specs https://pytorch.org/docs/stable/generated/torch.unique.html. Anyways the returned data type is also `int64`:
```
>>> torch.unique(t, return_counts=True)[1].dtype
torch.int64
```

File: docs/ops/movement/Unique_10.md (hunk 3 above)
Comment (tomdol) [reply docs/ops/movement/Unique_10.md:30]: PyTorch seems to sort the values by default while TF maintains the original order of the elements in the input (according to the docs):
```
>>> t
tensor([[3., 2., 1.],
        [1., 2., 3.],
        [4., 5., 6.]])
>>> torch.unique(t)
tensor([1., 2., 3., 4., 5., 6.])
>>> torch.unique(t,dim=0)
tensor([[1., 2., 3.],
        [3., 2., 1.],
        [4., 5., 6.]])
```

File: docs/ops/opset10.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
+# opset9 {#openvino_docs_ops_opset9}
```
Comment (mitruska) [inline docs/ops/opset10.md:1]:
```suggestion
# opset10 {#openvino_docs_ops_opset10}
```

File: docs/ops/opset10.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+This specification document describes the `opset9` operation set supported in OpenVINO™.
```
Comment (mitruska) [inline docs/ops/opset10.md:3]:
```suggestion
This specification document describes the `opset10` operation set supported in OpenVINO™.
```

File: docs/ops/opset10.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+Support for each particular operation from the list below depends on the capabilities of an inference plugin
+and may vary among different hardware platforms and devices. Examples of operation instances are provided as IR V10 xml
+snippets. Such IR is generated by the Model Optimizer. The semantics match corresponding nGraph operation classes
+declared in `namespace opset9`.
```
Comment (mitruska) [inline docs/ops/opset10.md:7]:
```suggestion
declared in `namespace opset10`.
```

File: docs/ops/movement/Unique_10.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
```
Comment (mitruska) [inline docs/ops/movement/Unique_10.md:1]:
```suggestion
# Unique  {#openvino_docs_ops_movement_Unique_10}
```

File: docs/ops/opset10.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+
+## Table of Contents <a name="toc"></a>
```
Comment (mitruska) [inline docs/ops/opset10.md:10]: Please add: 
- opset10 to: [docs/ops/opset.md](https://github.com/openvinotoolkit/openvino/pull/12162/commits/605d6d25a0a36b09b5b7a72714c9db8ff6b3f9f5)

- Unique op to: [/docs/OV_Runtime_UG/Operations_specifications.md](https://github.com/openvinotoolkit/openvino/blob/94ef65f0bd07221f79f5a57c7dbdc389eb2e09f9/docs/OV_Runtime_UG/Operations_specifications.md)

File: docs/ops/movement/Unique_10.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            <dim>3</dim>
+            <dim>-1</dim>
+        </port>
+        <port id="3" precision="I32">
+            <dim>-1</dim>
+        </port>
+        <port id="4" precision="I32">
+            <dim>-1</dim>
+        </port>
+        <port id="5" precision="I32">
```
Comment (mitruska) [inline docs/ops/movement/Unique_10.md:80]: In the current version of the specification, this output is always "I64"

File: docs/ops/movement/Unique_10.md [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        <port id="1" precision="FP32">
+            <dim>-1</dim>
+        </port>
+        <port id="2" precision="I32">
+            <dim>-1</dim>
+        </port>
+        <port id="3" precision="I32">
+            <dim>-1</dim>
+        </port>
+        <port id="4" precision="I32">
```
Comment (mitruska) [inline docs/ops/movement/Unique_10.md:106]: In the current version of the specification, this output is always "I64".

File: docs/ops/movement/Unique_10.md (hunk 2 above)
Comment (mitruska) [reply docs/ops/movement/Unique_10.md:43]: What do you think about an attribute to set the output type for the 4th output? 
Similar as *index_element_type* for indices (output 3).

Comment (tomdol) [reply docs/ops/movement/Unique_10.md:43]: > What do you think about an attribute to set the output type for the 4th output? Similar as _index_element_type_ for indices (output 3).

This is something I've considered as well, it's in the last bullet point in the PR's description.  Overall I don't see the need to configure the data type for this particular output, I think it should rather be an unsigned data type but ONNX requires `int64` - that's why I decided to go for this element type in OV. I'm open to suggestions though.

File: docs/ops/opset.md [hunk 12]
```diff
@@ -24,6 +25,7 @@ This topic provides a complete list of available sets of operations supported in
 
 | OpenVINO™ Version | Actual Operations Set            |
 | :---------------- | :------------------------------- |
+| 2022.3            | [opset9](opset10.md)   |
```
Comment (mitruska) [inline docs/ops/opset.md:28]:
```suggestion
| 2022.3            | [opset10](opset10.md)   |
```

File: docs/ops/movement/Unique_10.md (hunk 8 above)
Comment (mitruska) [reply docs/ops/movement/Unique_10.md:1]: This line should be aligned with the directory of the op category. Actually in this case rather the `Unique_10.md` should be moved from `/movement` to `/comparison` directory, instead of changing the `{#openvino_docs_ops_comparison_Unique_10}`.

File: docs/ops/movement/Unique_10.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+  When provided this input is used to "divide" the input tensor into slices along the specified axis before unique elements processing starts. When this input is not provided the operator works on a flattened version of the input tensor (elementwise processing).
+
+**Outputs**
+
+* **1**: The output tensor containing unique elements (individual values or subtensors). This tensor's type matches the type of the first input tensor: *T*. The values in this tensor are either sorted ascendingly or maintain the same order as in the input tensor. The shape of this output depends on the values of the input tensor and will very often be dynamic. Please refer to the article describing how [Dynamic Shapes](https://docs.openvino.ai/latest/openvino_docs_OV_UG_DynamicShapes.html) are handled in OpenVINO.
+* **2**: The output tensor containing indices of the locations of unique elements. The indices map the elements in the first output tensor to their locations in the input tensor. The index always points to the first occurence of a given unique output element in the input tensor. This is a 1D tensor with type *T_IND*.
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:41]: I think we should tell `This is a 1D tensor with type *index_element_type*.` and then *T_IND* is not needed. The next output is the same but I am not sure if we need a separate type attribute for it.

File: docs/ops/movement/Unique_10.md [hunk 14]
```diff
@@ -0,0 +1,111 @@
+# Unique  {#openvino_docs_ops_movement_Unique_10}
+
+**Versioned name**: *Unique-10*
+
+**Category**: *Data movement*
+
+**Short description**: *Unique* finds unique elements of a given input tensor. Depending on the operator's attributes it either looks for unique values globally in the whole (flattened) tensor or performs the search along the specified axis.
+
+**Detailed description**
+The operator can either work in elementwise mode searching for unique values in the whole tensor or it can consider slices of the input tensor along the specified axis. This way the op is able to find unique subtensors in the input data. Except for the unique elements the operator also produces the indices of the unique elements in the input tensor and the number of occurences of unique elements in the input data.
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:10]: occurences -> occurrences

File: docs/ops/movement/Unique_10.md [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+**Attributes**:
+
+* *sorted*
+
+  * **Description**: controls whether the unique elements in the output tensor are sorted in ascending order.
+  * **Range of values**:
+    * false - output tensor's elements are not sorted
+    * true - output tensor's element are sorted
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:19]: elements

File: docs/ops/movement/Unique_10.md (hunk 13 above)
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:41]: occurrence

File: docs/ops/movement/Unique_10.md [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+* **2**: The output tensor containing indices of the locations of unique elements. The indices map the elements in the first output tensor to their locations in the input tensor. The index always points to the first occurence of a given unique output element in the input tensor. This is a 1D tensor with type *T_IND*.
+* **3**: The output tensor containing indices of the locations of elements of the input tensor in the first output tensor. This means that for each element of the input tensor this output will point to the unique value in the first output tensor of this operator. This is a 1D tensor with type *T_IND*.
+* **4**: The output tensor containing the number of occurences of each unique value produced by this operator in the first output tensor. This is a 1D tensor with type `int64`.
```
Comment (rkazants) [inline docs/ops/movement/Unique_10.md:43]: occurrence

**Review by mitruska — APPROVED** [review]
Approved the current version, but:
- The discussion about an additional attribute to set the type of the last output should be finalized.
- The category for this op could be more suitable than "movement".

**Review by rkazants — APPROVED** [review]

**Comment by rkazants** [comment]
> Approved the current version, but:
> 
> * The discussion about an additional attribute to set the type of the last output should be finalized.
> * The category for this op could be more suitable than "movement".

@tomdol, @mitruska, maybe `condition` or `comparison` category would be better?

**Comment by tomdol** [comment]
> > Approved the current version, but:
> > 
> > * The discussion about an additional attribute to set the type of the last output should be finalized.
> > * The category for this op could be more suitable than "movement".
> 
> @tomdol, @mitruska, maybe `condition` or `comparison` category would be better?

I've considered the category too and from my perspective movement was the best because the Unique operation does something between Gather and Transpose. It selectively copies the data to the output tensor and optionally sorts it. Gather and Transpose are categorized as "movement" and given all that I thought that Unique could go here too.
